src/OrchardCore.Modules/OrchardCore.Queries/Sql/SqlQuerySource.cs

```diff
@@ -84,33 +106,29 @@ public async Task<IQueryResults> ExecuteQueryAsync(Query query, IDictionary<stri
                     else
                     {
                         column = rowDictionary.FirstOrDefault(kv => kv.Value is long).Key
-                            ?? rowDictionary.First().Key;
+                             ?? rowDictionary.First().Key;
                     }
                 }
 
-                return rowDictionary.TryGetValue(column, out var documentIdObject) && documentIdObject is long documentId
-                    ? documentId
-                    : 0;
-            }).ToArray();
-
+                return (rowDictionary.TryGetValue(column, out var documentIdObject) ? documentIdObject : null) switch
+                {
+                    long longValue => longValue,
+                    int intValue => intValue, 
+                    { } otherObject => 
+                        long.TryParse(otherObject.ToString(), CultureInfo.InvariantCulture, out var parsedValue)
+                            ? parsedValue
+                            : -1,
+                    _ => -1,
+                };
+            })
+            .Where(id => id >= 0)
+            .ToArray();
```

Comment:
I think the following version is more readable and utilize vertical code instead of horizontal. 
```suggestion
                if (rowDictionary.TryGetValue(column, out var documentIdObject))
                {
					return documentIdObject switch
	                {
                    	long longValue => longValue,
                    	int intValue => intValue, 
                    	{ } otherObject => 
                        	long.TryParse(otherObject.ToString(), CultureInfo.InvariantCulture, out var parsedValue)
                            ? parsedValue
                            : 0,
                    	_ => 0,
                	};
                }

                return 0;
            }).ToArray();
```

Reply:
Why is it okay to return 0 when the ID column is not found or can't be resolved? Isn't it safer to keep the `.Where(id => id > 0)`?

Reply:
> Why is it okay to return 0 when the ID column is not found or can't be resolved? Isn't it safer to keep the .Where(id => id > 0)?

@sarahelsaig I am just trying to avoid having the additional loop. Maybe then change this code to the following:
```
var documentIds = new List<long>();

foreach(var item in queryResult)
{
     var rowDictionary = (IDictionary<string, object>)row;

    if (column == null)
    {
        if (rowDictionary.ContainsKey(nameof(ContentItemIndex.DocumentId)))
        {
            column = nameof(ContentItemIndex.DocumentId);
        }
        else
        {
            column = rowDictionary.FirstOrDefault(kv => kv.Value is long).Key
            ?? rowDictionary.First().Key;
        }
    }

    if (!rowDictionary.TryGetValue(column, out var documentIdObject)) 
    {
        continue;
    }

    if (documentIdObject is long longDocumentId)
    {
        documentIds.Add(longDocumentId);
    } 
    else if (documentIdObject is int intDocumentId)
    {
        documentIds.Add(intDocumentId);
    } 
}
```

The above code removed the extra loop, and also is more readable.

Reply:
@MikeAlhayek Just to note, `Where` does not do an additional looping over the original list. This is all delayed until enumerating the final enumerator, in this case done by `ToArray()`. The original list is effectively enumerated only once in these cases.